### PR TITLE
[develop] Conan

### PR DESCRIPTION
**Pull request by scottarnette**
## Change Description

Added full support for using Conan to manage dependencies when building EOSIO:
- Buildkite pipeline functionality now matches pinned/unpinned builds. Supply `USE_CONAN=true` in Buildkite to run a build using Conan to manage dependencies.
- You can also provide `-DUSE_CONAN=true` to cmake when building locally to have Conan manage dependencies.
- Docs added in `.conan` directory to walkthrough steps of using Conan for each supported platform.
  - CI steps are dynamically generated from these documents.
  - Document added in `scripts` directory to detail build process and dependencies when not using Conan.
- Restructured `.cicd/platforms` to support Conan images.
  - Created partial Dockerfiles used for Conan builds.
  - Separate all platforms into pinned, unpinned, and conan folder.
  - Updated all references to pinned `IMAGE_TAG`s to include `-pinned` as a suffix.
- Updated `.cicd/generate-base-images.sh` to support use in Anka tag creation and generating Conan images.
  - Anka tag generation now calls this script instead of the Mac scripts directly.
  - Dynamically append to Conan Dockerfiles and Mac scripts based on Environment Steps section of Conan documentation per platform.
- Updated `.cicd/build.sh` to support Conan.
  - Dynamically generate build steps for Conan based on Build Steps section of Conan documentation.
- Updated `.cicd/test.sh` to support Conan.
  - Dynamically write the Dockerfile at this step again to ensure we grab the correct Docker image for testing Conan. Otherwise, the image would not exist.
- Updated `.cicd/package.sh` to support new platforms structure.
- Updated `.travis.yml` to support new `IMAGE_TAG` and `PLATFORM_TYPE` variables.

Other changes:
- Removed latest tag in Contracts Builder. This would flip-flop between being a image based off different branches and was unreliable.
- Updated `scripts/serial-test.sh` and `scripts/long-running-test.sh` to not start mongod if it does not exist.
  - Added bonus of no longer needing to install Mongo dependencies during Mac builds on Travis, speeding that build time up.
- Updated Ubuntu 18 unpinned to use Clang-7 to match AmazonLinux2 unpinned.

See:
https://buildkite.com/EOSIO/eosio/builds/18287 | Pinned
https://buildkite.com/EOSIO/eosio-build-unpinned/builds/760 | Unpinned
https://buildkite.com/EOSIO/eosio-conan/builds/458 | Conan
https://travis-ci.com/EOSIO/eos/builds/132238360 | Travis build (Mac failures are known.)
## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: .conan/conanfile-pinned.txt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+OpenSSL/latest_1.1.1x@conan/stable
+libcurl/7.66.0@bincrafters/stable
+libusb/1.0.23@bincrafters/stable
+gettext/0.20.1@bincrafters/stable
+gmp/6.1.2@bincrafters/stable
+zlib/1.2.11@conan/stable
+
+[options]
+libusb*:enable_udev=False
+gettext:shared=True
```
Comment (spoonincode) [inline .conan/conanfile-pinned.txt]: Doesn't this add a new dependency on Linux? (`gettext()` is just in libc for Linux) Is it possible to only enable a dep on macos vs linux?

Comment (NorseGaud) [reply .conan/conanfile-pinned.txt]: @ericiles Any ideas?

Comment (scottarnette) [reply .conan/conanfile-pinned.txt]: We could define separate conanfiles for Mac vs. Linux and have that logic in the CMake level. Let's discuss this more in tomorrow's meeting.

**Comment by spoonincode** [comment]
With #8118 merged, we can remove the gettext dep in here now

File: .conan/conanfile-pinned.txt (hunk 1 above)
Comment (scottarnette) [reply .conan/conanfile-pinned.txt]: Gettext removed in #8118. Commit 0f2f92c removes this dependency from the conanfiles.

**Comment by spoonincode** [comment]
I attempted to follow the instructions for a Ubuntu 18.04 build and failed twice in two different ways. On my first attempt (that was in a container that I've used previously) I would get an error when trying to compile gmp:
```
checking C++ compiler g++ -DNDEBUG -D_GLIBCXX_USE_CXX11_ABI=1 -m64 -O3 -fPIC -stdlib=libstdc++... no                                                                                                    
configure: error: C++ compiler not available, see config.log for details                                                                                                                                
gmp/6.1.2@bincrafters/stable:                                                                                                                                                                           
WARN: libcurl/7.66.0@bincrafters/stable: requirement zlib/1.2.11 overridden by your conanfile to zlib/1.2.11@conan/stable                                                                               
WARN: openssl/1.1.1d: requirement zlib/1.2.11 overridden by libcurl/7.66.0@bincrafters/stable to zlib/1.2.11@conan/stable                                                                               
gmp/6.1.2@bincrafters/stable: ERROR: Package '244633fabf6cb8b674d1d43097294d0f40cc421a' build failed                                                                                                    
gmp/6.1.2@bincrafters/stable: WARN: Build folder /home/ubuntu/.conan/data/gmp/6.1.2/bincrafters/stable/build/244633fabf6cb8b674d1d43097294d0f40cc421a                                                   
ERROR: gmp/6.1.2@bincrafters/stable: Error in build() method, line 66                                                                                                                                   
        autotools = self._configure_autotools()                                                                                                                                                         
while calling '_configure_autotools', line 62                                                                                                                                                           
        self._autotools.configure(args=configure_args, configure_dir=self._source_subfolder)                                                                                                            
        ConanException: Error 1 while executing source_subfolder/configure '--disable-assembly' '--enable-shared' '--disable-static' '--enable-cxx' '--prefix=/home/ubuntu/.conan/data/gmp/6.1.2/bincraf
ters/stable/package/244633fabf6cb8b674d1d43097294d0f40cc421a' '--bindir=${prefix}/bin' '--sbindir=${prefix}/bin' '--libexecdir=${prefix}/bin' '--libdir=${prefix}/lib' '--includedir=${prefix}/include' 
'--oldincludedir=${prefix}/include' '--datarootdir=${prefix}/share'                                                                                                                                     
CMake Error at .conan/conan.cmake:399 (message):                                                                                                                                                        
  Conan install failed='1'                                                                                                                                                                              
Call Stack (most recent call first):                                                                                                                                                                    
  .conan/conan.cmake:485 (conan_cmake_install)                                                                                                                                                          
  CMakeLists.txt:52 (conan_cmake_run)                  
```
Then I started from a fresh container and performed the following
```
# just update everything to the latest first
sudo apt-get update
sudo apt-get upgrade
# now follow the instructions (more or less; the instructions don't say you need to run some stuff as root, or that installing cmake to /usr/local/cmake needs to have it now in its path)
$ sudo apt-get install -y clang-7 llvm-7-dev python3 python3-dev python3-pip git curl automake jq python2.7 python2.7-dev
$ sudo pip3 install conan
$ curl -LO https://github.com/Kitware/CMake/releases/download/v3.15.3/cmake-3.15.3-Linux-x86_64.sh
$ sudo mkdir -p /usr/local/cmake
$ chmod a+x cmake-3.15.3-Linux-x86_64.sh 
$ sudo ./cmake-3.15.3-Linux-x86_64.sh --prefix=/usr/local/cmake --skip-license
$ PATH=/usr/local/cmake/bin/:$PATH
$ git clone -b conan-poc-v2 --recursive https://github.com/EOSIO/eos
$ cd eos
$ cmake -DCMAKE_BUILD_TYPE='Release' -DPKG_CONFIG_USE_CMAKE_PREFIX_PATH=ON -DCMAKE_CXX_COMPILER='clang++-7' -DCMAKE_C_COMPILER='clang-7' -DLLVM_DIR='/usr/lib/llvm-7/lib/cmake/llvm' -DUSE_CONAN=true -Bbuild
```
But this failed with 
```
- Conan: checking conan executable in path                                                                                                                                                             
-- Conan: Found program /usr/local/bin/conan                                                                                                                                                            
-- Conan: Version found Traceback (most recent call last):                                                                                                                                              
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 574, in _build_master                                                                                                           
    ws.require(__requires__)                                                                                                                                                                            
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 892, in require                                                                                                                 
    needed = self.resolve(parse_requirements(requirements))                                                                                                                                             
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 783, in resolve                                                                                                                 
    raise VersionConflict(dist, req).with_context(dependent_req)                                                                                                                                        
pkg_resources.ContextualVersionConflict: (six 1.11.0 (/usr/lib/python3/dist-packages), Requirement.parse('six==1.12'), {'astroid'})                                                                     
                                                                                                                                                                                                        
During handling of the above exception, another exception occurred:                                                                                                                                     
                                                                                                                                                                                                        
Traceback (most recent call last):
  File "/usr/local/bin/conan", line 6, in <module> 
    from pkg_resources import load_entry_point
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3088, in <module>
    @_call_aside
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3072, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3101, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 576, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 589, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 783, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (six 1.11.0 (/usr/lib/python3/dist-packages), Requirement.parse('six==1.12'), {'astroid'})

CMake Error at .conan/conan.cmake:530 (if):
  if given arguments:

    "VERSION_LESS" "1.0.0"

  Unknown arguments specified
Call Stack (most recent call first):
  CMakeLists.txt:41 (conan_check)
```

**Comment by scottarnette** [comment]
Thanks for the feedback @spoonincode . I'll investigate this more. 

As an update, I'm splitting this PR into 2 smaller PRs; one for the Conan specific changes and the second for the supporting CI changes.

**Pull request closed without merging**